File: 常见的加密算法/DES.js
// 引入CryptoJS库，实际使用时需要通过script标签引入
// <script src="https://cdnjs.cloudflare.com/ajax/libs/crypto-js/4.1.1/crypto-js.min.js"></script>
const CryptoJS = require('crypto-js')

/**
 * DES加密方法
 * @param {string} key 加密密钥，必须是8字节
 * @param {string} data 待加密的数据
 * @returns {string} 加密后的Base64字符串
 */
function desEncrypt(key, data) {
    try {
        // 验证密钥长度
        if (key.length !== 8) {
            throw new Error("DES密钥必须是8字节长度");
        }

        // 将密钥转换为WordArray
        const keyHex = CryptoJS.enc.Utf8.parse(key);

        // 加密，使用ECB模式和Pkcs7填充
        const encrypted = CryptoJS.DES.encrypt(data, keyHex, {
            mode: CryptoJS.mode.ECB,
            padding: CryptoJS.pad.Pkcs7
        });

        // 返回Base64格式的加密结果
        return encrypted.toString();
    } catch (e) {
        return `加密失败: ${e.message}`;
    }
}

/**
 * DES解密方法
 * @param {string} key 解密密钥，必须是8字节
 * @param {string} encryptedData 待解密的Base64字符串
 * @returns {string} 解密后的原始数据
 */
function desDecrypt(key, encryptedData) {
    try {
        // 验证密钥长度
        if (key.length !== 8) {
            throw new Error("DES密钥必须是8字节长度");
        }

        // 将密钥转换为WordArray
        const keyHex = CryptoJS.enc.Utf8.parse(key);

        // 解密
        const decrypted = CryptoJS.DES.decrypt(encryptedData, keyHex, {
            mode: CryptoJS.mode.ECB,
            padding: CryptoJS.pad.Pkcs7
        });

        // 将解密结果转换为UTF-8字符串
        return decrypted.toString(CryptoJS.enc.Utf8);
    } catch (e) {
        return `解密失败: ${e.message}`;
    }
}

/**
 * 主函数，演示DES加密解密流程
 */
function main() {
    // 密钥必须是8字节
    const key = "12345678";

    // 原始数据
    const originalData = "这是一个DES加密解密的测试案例";
    console.log(`原始数据: ${originalData}`);

    // 加密
    const encrypted = desEncrypt(key, originalData);
    console.log(`加密后: ${encrypted}`);

    // 解密
    const decrypted = desDecrypt(key, encrypted);
    console.log(`解密后: ${decrypted}`);
}

// 执行主函数
main();

File: 常见的加密算法/DES.py
from Crypto.Cipher import DES
from Crypto.Util.Padding import pad, unpad
import base64


def des_encrypt(key, data):
    """
    使用DES算法加密数据
    :param key: 加密密钥，必须是8字节
    :param data: 待加密的数据
    :return: 加密后的Base64字符串
    """
    try:
        # 确保密钥是8字节
        if len(key) != 8:
            raise ValueError("DES密钥必须是8字节长度")

        # 创建DES实例，使用ECB模式
        cipher = DES.new(key.encode('utf-8'), DES.MODE_ECB)

        # 对数据进行填充，使其长度为8的倍数
        padded_data = pad(data.encode('utf-8'), DES.block_size)

        # 加密并进行Base64编码
        encrypted_data = cipher.encrypt(padded_data)
        return base64.b64encode(encrypted_data).decode('utf-8')
    except Exception as e:
        return f"加密失败: {str(e)}"


def des_decrypt(key, encrypted_data):
    """
    使用DES算法解密数据
    :param key: 解密密钥，必须是8字节
    :param encrypted_data: 待解密的Base64字符串
    :return: 解密后的原始数据
    """
    try:
        # 确保密钥是8字节
        if len(key) != 8:
            raise ValueError("DES密钥必须是8字节长度")

        # 创建DES实例，使用ECB模式
        cipher = DES.new(key.encode('utf-8'), DES.MODE_ECB)

        # 对Base64编码的数据进行解码
        decoded_data = base64.b64decode(encrypted_data)

        # 解密并去除填充
        decrypted_data = unpad(cipher.decrypt(decoded_data), DES.block_size)
        return decrypted_data.decode('utf-8')
    except Exception as e:
        return f"解密失败: {str(e)}"


if __name__ == "__main__":
    # 密钥必须是8字节
    key = "12345678"

    # 原始数据
    original_data = "这是一个DES加密解密的测试案例"
    print(f"原始数据: {original_data}")

    # 加密
    encrypted = des_encrypt(key, original_data)
    print(f"加密后: {encrypted}")

    # 解密
    decrypted = des_decrypt(key, encrypted)
    print(f"解密后: {decrypted}")
